spring-integration-core/src/main/java/org/springframework/integration/util/MessagingMethodInvokerHelper.java

```diff
@@ -277,13 +289,15 @@ private void prepareEvaluationContext() {
 			}
 		}
 		else {
-			AnnotatedMethodFilter filter = new AnnotatedMethodFilter(this.annotationType, this.methodName,
-					this.requiresReply);
-			Assert.state(canReturnExpectedType(filter, targetType, context.getTypeConverter()),
+			AnnotatedMethodFilter annotatedMethodFilter = new AnnotatedMethodFilter(this.annotationType,
+					this.methodName, this.requiresReply);
+			Assert.state(canReturnExpectedType(annotatedMethodFilter, targetType, context.getTypeConverter()),
 					"Cannot convert to expected type (" + this.expectedType + ") from " + this.method);
-			context.registerMethodFilter(targetType, filter);
+			context.registerMethodFilter(targetType, annotatedMethodFilter);
+			declaredMethodResolver.registerMethodFilter(targetType, annotatedMethodFilter);
 		}
 		context.setVariable("target", this.targetObject);
+		context.addMethodResolver(declaredMethodResolver);
```

Comment:
Needs something like...

	context.addMethodResolver(declaredMethodResolver);
	if (methodResolvers.size() == 2 && methodResolvers.get(1).getClass().equals(ReflectiveMethodResolver.class)) {
		methodResolvers.remove(methodResolvers.get(1));
	}

Otherwise, caching is disabled - see `MethodReference` line 175.